File: android/app/src/main/java/com/marnock/app/transport/Framing.kt
package com.marnock.app.transport

import java.nio.ByteBuffer
import java.nio.ByteOrder

object Framing {
    fun encode(jsonUtf8: ByteArray): ByteArray {
        val buf = ByteBuffer.allocate(4 + jsonUtf8.size).order(ByteOrder.BIG_ENDIAN)
        buf.putInt(jsonUtf8.size)
        buf.put(jsonUtf8)
        return buf.array()
    }

    fun decode(frame: ByteArray): ByteArray {
        if (frame.size < 4) return frame
        val n = ByteBuffer.wrap(frame, 0, 4).order(ByteOrder.BIG_ENDIAN).int
        if (n in 0 until frame.size && n + 4 == frame.size) {
            return frame.copyOfRange(4, frame.size)
        }
        // Also accept raw JSON text frames without length prefix.
        return frame
    }
}

File: macos/Marnock/Marnock/Transport/Framing.swift
import Foundation

enum Framing {
    static func encode(_ jsonUTF8: Data) -> Data {
        var length = UInt32(jsonUTF8.count).bigEndian
        var out = Data(bytes: &length, count: 4)
        out.append(jsonUTF8)
        return out
    }

    static func decode(_ frame: Data) -> Data {
        guard frame.count >= 4 else { return frame }
        let n = frame.prefix(4).withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        if Int(n) + 4 == frame.count {
            return frame.dropFirst(4)
        }
        return frame
    }
}
